File: Array&Hashing/1733. Minimum Number of People to Teach/solution.py
class Solution:
    def minimumTeachings(self, n: int, languages: List[List[int]], friendships: List[List[int]]) -> int:

        langs = [set(lang) for lang in languages]
        need_to_teach = []
        for u, v in friendships:
            i, j = u-1, v-1
            
            if len(langs[i] & langs[j])>0:
                continue
            need_to_teach.append([u,v])
        
        res = inf
        for teached_lang in range(1, n+1):
            learned = langs.copy()
            teached = 0

            for u, v in need_to_teach:
                i, j = u-1, v-1
                
                if teached_lang not in learned[i]:
                    learned[i].add(teached_lang)
                    teached += 1
                if teached_lang not in learned[j]:
                    learned[j].add(teached_lang)
                    teached += 1
            
            res = min(res, teached)
        return res

File: Array&Hashing/1733. Minimum Number of People to Teach/solution.cpp
class Solution {
public:
    int minimumTeachings(int n, vector<vector<int>>& languages, vector<vector<int>>& friendships) {
        int m = languages.size();

        unordered_map<int, unordered_set<int>> lang2people;
        for (int i=1; i<=m; i++) {
            for (int lang : languages[i-1]) {
                lang2people[lang].insert(i);
            }
        }

        vector<int> need_learn(m+1, false);
        for (auto& friendship : friendships) {
            bool can_communicate = false;
            for (int lang = 1; lang <= n; lang++) {
                if (lang2people[lang].count(friendship[0]) && lang2people[lang].count(friendship[1])) {
                    can_communicate = true;
                    break;
                }
            }

            if (!can_communicate) {
                need_learn[friendship[0]] = true;
                need_learn[friendship[1]] = true;
            }
        }


        int res = m;
        for (int lang=1; lang<=n; lang++) {
            int teached = 0;
            
            for (int i=1; i<=m; i++) {
                if (need_learn[i] && !lang2people[lang].count(i)) teached++;
            }
            res = min(res, teached);
        }
        return res;
    }
};